src/main/java/com/conveyal/gtfs/stats/StopStats.java

```diff
@@ -1,7 +1,239 @@
+        stats = fs;
+        routeStats = stats.route;
+    }
+
+    public Set<Route> getRoutes (String stop_id) {
+        return feed.patterns.values().stream()
+                .filter(p -> p.orderedStops.contains(stop_id))
+                .map(p -> feed.routes.get(p.route_id))
+                .collect(Collectors.toSet());
+    }
+
+    public int getRouteCount (String stop_id) {
+        return getRoutes(stop_id).size();
+    }
+
+    public int getTripCountForDate (String stop_id, LocalDate date) {
+        return (int) getTripsForDate(stop_id, date).size();
+    }
+
+    /** Get list of trips for specified date of service */
+    public List<Trip> getTripsForDate (String stop_id, LocalDate date) {
+        List<String> tripIds = stats.getTripsForDate(date).stream()
+                .map(trip -> trip.trip_id)
+                .collect(Collectors.toList());
+
+        return feed.getStopTimesForStop(stop_id).stream()
+                .map(t -> feed.trips.get(t.b.a)) // map to trip ids
+                .filter(t -> tripIds.contains(t.trip_id)) // filter by trip_id list for date
+                .collect(Collectors.toList());
+    }
+
+    public int getAverageHeadwayForStop (String stop_id, LocalDate date, LocalTime from, LocalTime to) {
+        List<Trip> tripsForStop = feed.getStopTimesForStop(stop_id).stream()
+                .map(t -> feed.trips.get(t.b.a))
+                .filter(trip -> feed.services.get(trip.service_id).activeOn(date))
+                .collect(Collectors.toList());
+
+        return getStopHeadwayForTrips(stop_id, tripsForStop, from, to);
+    }
+
+    /** Get the route headway for a given service date at a stop over a time window, in seconds */
+    public Map<String, Integer> getRouteHeadwaysForStop (String stop_id, LocalDate date, LocalTime from, LocalTime to) {
+        Map<String, Integer> routeHeadwayMap = new HashMap<>();
+        List<Route> routes = feed.patterns.values().stream()
+                .filter(p -> p.orderedStops.contains(stop_id))
+                .map(p -> feed.routes.get(p.route_id))
+                .collect(Collectors.toList());
+
+        for (Route route : routes) {
+            routeHeadwayMap.put(route.route_id, getHeadwayForStopByRoute(stop_id, route.route_id, date, from, to));
+        }
+        return routeHeadwayMap;
+    }
+
+    /** Get the average headway for a set of trips at a stop over a time window, in seconds */
+    public int getStopHeadwayForTrips (String stop_id, List<Trip> trips, LocalTime from, LocalTime to) {
+        TIntList timesAtStop = new TIntArrayList();
+
+        for (Trip trip : trips) {
+            StopTime st;
+            try {
+                // use interpolated times in case our common stop is not a time point
+                st = StreamSupport.stream(feed.getInterpolatedStopTimesForTrip(trip.trip_id).spliterator(), false)
+                        .filter(candidate -> candidate.stop_id.equals(stop_id))
+                        .findFirst()
```

Comment:
This is looking at the first stop not the requested stop, which is going to mess up when you have multiple terminals.